feat(useLocalStorage): Initial implementation of useLocalStorage

palexcast opened the pull request:
This addresses https://github.com/nartc/ngxtension-platform/issues/245

Adds a utility that synchronizes localStorage changes across tabs, and has an optional validate method to verify the validity of the object.

I can create a PR for a similar/same thing for sessionStorage if this seems like a good idea/implementation.
Would love some feedback before I do that though :)

ilirbeqirii commented:
Should this utility function also, follow the "inject***" naming pattern like other functions? @eneajaho

palexcast commented:
> Should this utility function also, follow the "inject***" naming pattern like other functions? @eneajaho

I've got no strong opinion on the naming of this, I can fix it if you want it to change :) 

I also have some stuff I want to clean up, like passing the validate func as a part of an optional "options" param, as there might be other stuff to add, and it fits the "signal" way of passing options.

eneajaho commented:
Hello @palexcast 
Thanks for the PR.

Wanted to link this other implementation here https://twitter.com/_wallstreetdev/status/1767693809503084873 

Would like to start a conversation about these implementations and find the best way to go forward.

guzmanoj commented:
Hello WSD here :)
Disclaimer: I just liked a lot [Mike Hartington's](https://twitter.com/mhartington/status/1765132445895274635) idea but my code is also amateurish. I'm just starting with signals :)
In terms of the implementation it really depends on the direction you guys want to take. The approach I've shared is heavily inspired by React [use-local-storage-state](https://www.npmjs.com/package/use-local-storage-state) and is 100% based on Signals. It also lacks of In-Memory-Data support and proper error handling. This is all I have so far:

```
import { effect, signal, WritableSignal } from '@angular/core';

export interface LocalStorageOptions<T> {
  defaultValue?: T | (() => T);
  storageSync?: boolean;
  serializer?: {
    stringify: (value: unknown) => string;
    parse: (value: string) => unknown;
  };
}

function isFunction(value: unknown): value is (...args: unknown[]) => unknown {
  return typeof value === 'function';
}

function goodTry<T>(tryFn: () => T): T | undefined {
  try {
    return tryFn();
  } catch {
    return undefined;
  }
}

function parseJSON(value: string): unknown {
  return value === 'undefined' ? undefined : JSON.parse(value);
}

export function localStorageSignal<T>(key: string, options?: LocalStorageOptions<T>): WritableSignal<T | undefined> {
  const defaultValue = isFunction(options?.defaultValue) ? options.defaultValue() : options?.defaultValue;
  const serializer = options?.serializer || {
    stringify: JSON.stringify,
    parse: parseJSON
  };
  const storageSync = options?.storageSync ?? true;
  const initialStoredValue = goodTry(() => localStorage.getItem(key));
  const initialValue = initialStoredValue ? goodTry(() => serializer.parse(initialStoredValue) as T) : defaultValue;
  const internalSignal = signal<T | undefined>(initialValue);

  effect(() => {
    const value = internalSignal();
    if (value === undefined) {
      goodTry(() => localStorage.removeItem(key));
    } else {
      goodTry(() => localStorage.setItem(key, serializer.stringify(value)));
    }
  });

  if (storageSync) {
    const onStorage = (event: StorageEvent) => {
      if (event.storageArea === localStorage && event.key === key) {
        const newValue = event.newValue !== null ? (serializer.parse(event.newValue) as T) : undefined;
        internalSignal.set(newValue);
      }
    };
    window.addEventListener('storage', onStorage);
    effect(onCleanup => {
      onCleanup(() => window.removeEventListener('storage', onStorage));
    });
  }

  return internalSignal;
}

```

palexcast commented:
I'll take a look at both those implementations and see if I can adjust some of my implementation with some input from those.

If there is any more input/wishes regarding the direction of the implementation that would be greatly appreciated :)

palexcast commented:
@guzmanoj So I finally had some time, and the more I look at your implementation I really like it. It's clean and functional. 
I really don't feel comfortable just copy-pasting  it and taking credit for it though, the credit should be yours ;)

For now I copy-pasted your solution, and renamed it to `injectLocalStorage` like @ilirbeqirii mentioned, as it does adhere more to the convention of the lib.

In regards to error handling, I'm not sure about what the "signal" convention is regarding this.
Should `goodTry` throw an error, or just return `undefined` as the current implementation does.
Maybe @eneajaho has some thoughts?

My current thought regarding this, is that the `parse` function can be used to handle validation/errors. 
One of the tests demonstrate this.

I also unwrapped the serializer so you don't need to pass both `parse` and `stringify`. Adding docs instead should be enough to explain that they act as serializers. (just my opinion).

I've also adjusted the tests, but want to add some more, would be nice to get some feedback before moving on though.

eneajaho commented on libs/ngxtension/inject-local-storage/src/inject-local-storage.ts:
```diff
@@ -0,0 +1,69 @@
+import { effect, signal, type WritableSignal } from '@angular/core';
+
+export type LocalStorageOptions<T> = {
+	defaultValue?: T | (() => T);
```
Can we add some comments for each option here? 

It would make it easier when passing these options and to read some kind of comments on what they do and how they modify the way this function works.

palexcast replied:
I agree, added some comments :)

eneajaho commented:
Can you also update the PR description, so it is the same as the updated code we have?

palexcast commented on libs/ngxtension/inject-local-storage/src/inject-local-storage.ts:
```diff
@@ -0,0 +1,69 @@
+	const parse = isFunction(options?.parse) ? options?.parse : parseJSON;
+
+	const storageSync = options?.storageSync ?? true;
+	const initialStoredValue = goodTry(() => localStorage.getItem(key));
```
Fixed, nice catch! Didn't think about this :D

palexcast commented:
@eneajaho Updated the PR in regards to your comments :D

nartc commented on libs/ngxtension/inject-local-storage/src/inject-local-storage.ts:
```diff
@@ -0,0 +1,113 @@
+	return value === 'undefined' ? undefined : JSON.parse(value);
+}
+
+export const injectLocalStorage = <T>(
```
nit: please use `assertInjector()` to ensure this function is being called in an Injection Context. Also, can we go with top-level `function` declaration instead of Arrow Fn expression?

eneajaho commented:
Thanks for this @palexcast 🙌

palexcast commented:
Thanks for the reviews and help getting this merged @eneajaho :D 
First 'real' contribution to OS, so greatly appreciated!
